File: inst/d3/scatterplot/flower.js
// Center coordinates (no re-declaration of width/height)
const centerX = width / 2;
const centerY = height / 2;

// Flower configuration from options
const petalCount = options.petalCount || 5;
const petalLength = options.petalLength || 100;
const petalWidth = options.petalWidth || 50;
const petalColor = options.petalColor || "pink";
const petalStroke = options.petalStroke || "red";
const centerColor = options.centerColor || "yellow";
const centerStroke = options.centerStroke || "orange";
const centerRadius = options.centerRadius || 20;
const bgcol = options.bgcol || "white";
const centerText = options.centerText || "";

// Set background color
svg.style("background-color", bgcol);

// Petal group centered in the middle of the canvas
const petalGroup = svg.append("g")
  .attr("transform", `translate(${centerX}, ${centerY})`);

// Draw petals
for (let i = 0; i < petalCount; i++) {
  const angle = (360 / petalCount) * i;

  petalGroup.append("ellipse")
    .attr("cx", 0)
    .attr("cy", -petalLength / 2)
    .attr("rx", petalWidth / 2)
    .attr("ry", petalLength / 2)
    .attr("fill", petalColor)
    .attr("stroke", petalStroke)
    .attr("stroke-width", 2)
    .attr("transform", `rotate(${angle})`);
}

// Draw flower center
svg.append("circle")
  .attr("cx", centerX)
  .attr("cy", centerY)
  .attr("r", centerRadius)
  .attr("fill", centerColor)
  .attr("stroke", centerStroke)
  .attr("stroke-width", 2);



// Add the text in the center
svg.append("text")
  .attr("x", centerX)
  .attr("y", centerY)
  .attr("text-anchor", "middle")
  .attr("dominant-baseline", "middle")
  .style("font-size", options.centerTextSize || "16px")
  .style("fill", options.centerTextColor || "black")
  .style("font-family", options.font || "sans-serif")
  .text(centerText);

const rotationSpeed = options.rotationSpeed || 0
// Function to continuously rotate the petals around the center
function rotatePetals(angle = 0) {
  petalGroup
    .attr("transform", `translate(${centerX}, ${centerY}) rotate(${angle})`);

  d3.timeout(() => {
    rotatePetals((angle + rotationSpeed) % 360); // Adjust speed here
  }, 20); // 50 frames per second
}

// Start rotating
rotatePetals();

File: R/flower.R
#' Display a Rotating Flower Visualization
#'
#' @param petalCount The number of petals. Defaults to 6.
#' @param petalLength The length of each petal. Defaults to 100.
#' @param petalWidth The width of each petal. Defaults to 60.
#' @param petalColor The fill color of the petals. Defaults to "lightpink".
#' @param petalStroke The stroke color of the petals. Defaults to "deeppink".
#' @param centerRadius The radius of the flower's center circle. Defaults to 20.
#' @param centerColor The fill color of the center. Defaults to "gold".
#' @param centerStroke The stroke color of the center. Defaults to "darkorange".
#' @param centerText Optional. Text to display inside the center (e.g., a number or emoji).
#' @param centerTextSize The size of the center text. Defaults to 16.
#' @param centerTextColor The color of the center text. Defaults to "black".
#' @param font The font family for the center text. Defaults to "Verdana, Geneva, Tahoma, sans-serif".
#' @param bgcol The background color of the visualization. Defaults to "white".
#' @param rotationSpeed The speed of rotation (degrees per animation frame). Defaults to 2.
#' @param width The width of the SVG output. Optional.
#' @param height The height of the SVG output. Optional.
#'
#' @return An animated rotating flower SVG.
#' @export
#'
#' @examples
#' flower(
#'   petalCount = 5,
#'   petalColor = "plum",
#'   rotationSpeed = 1.5
#' )
flower <- function(
    petalCount = 6,
    petalLength = 100,
    petalWidth = 60,
    petalColor = "lightpink",
    petalStroke = "deeppink",
    centerRadius = 20,
    centerColor = "gold",
    centerStroke = "darkorange",
    centerText = NULL,
    centerTextSize = 16,
    centerTextColor = "black",
    font = "Verdana, Geneva, Tahoma, sans-serif",
    bgcol = "white",
    rotationSpeed = 2,
    width = NULL,
    height = NULL
) {
  if (grepl(";", font)) {
    stop("Please remove the ';' character from your font argument.")
  }

  r2d3::r2d3(
    data = NULL,
    script = system.file("d3/scatterplot/flower.js", package = "ddplot"),
    options = list(
      petalCount = petalCount,
      petalLength = petalLength,
      petalWidth = petalWidth,
      petalColor = petalColor,
      petalStroke = petalStroke,
      centerRadius = centerRadius,
      centerColor = centerColor,
      centerStroke = centerStroke,
      centerText = centerText,
      centerTextSize = centerTextSize,
      centerTextColor = centerTextColor,
      font = font,
      bgcol = bgcol,
      rotationSpeed = rotationSpeed,
      width = width,
      height = height
    )
  )
}
